### ISSUE TITLE
[API] New api in IMethodBinding to get the method JVM method signature

### ISSUE THREAD

**Issue by gayanper** [eclipse-jdt/eclipse.jdt.core#223]
Today the jvm method signature is part of the `getKey` method. But it is not guaranteed as it seems and its not the method signature only.

When using AST for java debug breakpoint evaluations, we need to get the method signature from the AST method bindings so it could be matched later with JDI method signatures.

One usage case for this API is searching for method declaration of the given line number and use the signature of that method to fine JDI locations in vscode java debugger plugin.

**Comment by gayanper** [comment]
@testforstephen please add more details if missing something. @jarthana @stephan-herrmann Do you think this is a good API candidate or Do you see another way to find the signature using JDT Core APIs.

**Comment by stephan-herrmann** [comment]
> Today the jvm method signature is part of the getKey method. But it is not guaranteed as it seems and its not the method signature only.

Right, `getKey()` explicitly does not guarantee a specific format.

> we need to get the method signature from the AST method bindings so it could be matched later with JDI method signatures.

`getKey()` is defined on `IBinding`, would you add smth like `getBytecodeSignature()` only on `IMethod`?

As the use case is debugging, I assume you would need erased signatures (`o.e.j.i.compiler.lookup.MethodBinding.signature(ClassFile)` rather than `o.e.j.i.compiler.lookup.MethodBinding.genericSignature()`), right?

Technically, I'm not sure that all information needed for such API is actually available on compiler bindings constructed on behalf of DOM AST.  You'd have to check if `o.e.j.i.compiler.lookup.TypeBinding.constantPoolName()` is guaranteed to be called on all types in that scenario, because after AST parsing finished, you may no longer be able to invoke it due to missing Scopes. This could go wrong in scenarios where the compiler is not instructed to generate byte code.

In particular, local types, anonymous types may only get their names very late in the game, as those names depend on things that are not visible in source code / AST.

I'm not even 100% sure that computation of those names is totally deterministic. There *is* a connection between type inference (which makes no guarantee about processing order) and management of LocalTypeBinding, which *might* influence name computation, see the story of `o.e.j.i.compiler.ast.LambdaExpression.LocalTypeSubstitutor`, for some unpleasant complexity.

**Comment by gayanper** [comment]
Thanks @stephan-herrmann I will have a look. I think for debug scenario we may not need to generic signature even though the JDI provided generic signature as well if its a generic method.

**Comment by stephan-herrmann** [comment]
Look what I just found: `org.eclipse.jdt.core.util.ExternalAnnotationUtil.extractGenericSignature(IMethodBinding)` :)

From there, looking left and right, the canonical idiom might be (all is API):
```
new BindingKey(methodBinding.getKey()).toSignature()
```

Can you try that? I can see that `BindingKey` is intended for keys from Java Model, not DOM AST, but in ExternalAnnotationUtil it seems to work fine also for keys obtaind from DOM bindings. Hm?
